TF_STATE_PERSIST_INTERVAL with invalid value causes "OPENTOFU CRASH" instead of user-friendly error

## Community note

> [!TIP]
> 👋 Hi there, OpenTofu community! The OpenTofu team prioritizes issues based on upvotes. Please make sure to upvote this issue and describe how it affects you in detail in the comments to show your support.

## OpenTofu Version

```shell
$ tofu version
OpenTofu v1.11.6
on linux_amd64
```

(Reproducible on current `main` as of v1.13 development period)

## OpenTofu Configuration Files

```hcl
# Any minimal configuration is sufficient to reproduce, e.g.:
terraform {
  required_providers {}
}

resource "null_resource" "example" {}
```

## Debug Output

Not required — the crash is immediately visible without `TF_LOG=trace`.

## Expected Behavior

When `TF_STATE_PERSIST_INTERVAL` is set to an invalid value (non-integer string, or an integer below the minimum of 20), `tofu apply` should print a clear, actionable error message and exit with a non-zero status code — the same way other configuration errors are reported.

Example of expected output:
```
╷
│ Error: Invalid value for TF_STATE_PERSIST_INTERVAL
│
│ The environment variable TF_STATE_PERSIST_INTERVAL must be an integer >= 20.
│ Got: "abc"
╵
```

## Actual Behavior

`tofu apply` crashes with an "OPENTOFU CRASH" message and a full goroutine stack trace, which is alarming and unhelpful for a simple user configuration mistake.

**Case 1 — non-integer value:**
```
TF_STATE_PERSIST_INTERVAL=abc tofu apply
```
```
OPENTOFU CRASH
...
panic: Can't parse value 'abc' of environment variable 'TF_STATE_PERSIST_INTERVAL'

goroutine 1 [running]:
runtime/debug.Stack(...)
...
```

**Case 2 — integer below minimum (20):**
```
TF_STATE_PERSIST_INTERVAL=5 tofu apply
```
```
OPENTOFU CRASH
...
panic: Can't use value lower than 20 for env variable TF_STATE_PERSIST_INTERVAL, got 5

goroutine 1 [running]:
...
```

## Steps to Reproduce

1. Create a minimal OpenTofu configuration (any `resource` block works).
2. Run `tofu init`.
3. Run one of the following:
   - `TF_STATE_PERSIST_INTERVAL=abc tofu apply`    — triggers panic on parse failure
   - `TF_STATE_PERSIST_INTERVAL=5 tofu apply`      — triggers panic on below-minimum check
   - `TF_STATE_PERSIST_INTERVAL=abc tofu destroy`  — same panic, `tofu destroy` also uses the same code path (`opApply`)
4. Observe the "OPENTOFU CRASH" output.

Note: `tofu plan` and `tofu refresh` are **not affected** — `getEnvAsInt` is only called in the apply code path.

## Root Cause (for reference)

`internal/backend/local/backend_apply.go`:

```go
// line 37-46
func getEnvAsInt(envName string, defaultValue int) int {
    if val, exists := os.LookupEnv(envName); exists {
        parsedVal, err := strconv.Atoi(val)
        if err == nil {
            return parsedVal
        }
        panic(fmt.Sprintf("Can't parse value '%s' of environment variable '%s'", val, envName))
    }
    return defaultValue
}

// line 115-119
persistInterval := getEnvAsInt(persistIntervalEnvironmentVariableName, defaultPersistInterval)
if persistInterval < defaultPersistInterval {
    panic(fmt.Sprintf("Can't use value lower than %d for env variable %s, got %d",
        defaultPersistInterval, persistIntervalEnvironmentVariableName, persistInterval))
}
```

Both `panic()` calls are triggered by ordinary user input errors. Although a `panicHandler` wraps the goroutine and prevents a raw Go crash, it still emits the "OPENTOFU CRASH" message — appropriate for unexpected internal errors, but not for user mistakes.

The fix is to return a `tfdiags.Diagnostics` error from `opApply` instead of calling `panic()`.

## Additional Context

- The `panicHandler` in `internal/logging/panic.go` intentionally catches panics and formats them as "OPENTOFU CRASH" for unexpected bugs. It is not intended to be used as a substitute for proper input validation.
- `TF_STATE_PERSIST_INTERVAL` allows users to tune state persistence frequency. Any user who misconfigures this variable will see a crash-level message, which is poor UX.
- This is a "fast path" candidate: small, self-contained fix with clear expected behavior.
- I'd be happy to submit a PR to fix this if the issue is confirmed.

## References

- `internal/backend/local/backend_apply.go` lines 37–46, 115–119
- `internal/logging/panic.go` (`panicHandler`)

--- a/internal/backend/local/backend_apply.go
+++ b/internal/backend/local/backend_apply.go
@@ -34,15 +34,21 @@ const (
 	persistIntervalEnvironmentVariableName = "TF_STATE_PERSIST_INTERVAL"
 )
 
-func getEnvAsInt(envName string, defaultValue int) int {
+func getEnvAsInt(envName string, defaultValue int) (int, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
 	if val, exists := os.LookupEnv(envName); exists {
 		parsedVal, err := strconv.Atoi(val)
-		if err == nil {
-			return parsedVal
+		if err != nil {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Invalid environment variable value",
+				fmt.Sprintf("The environment variable %s must be a valid integer, got %q.", envName, val),
+			))
+			return 0, diags
 		}
-		panic(fmt.Sprintf("Can't parse value '%s' of environment variable '%s'", val, envName))
+		return parsedVal, diags
 	}
-	return defaultValue
+	return defaultValue, diags
 }
 
 func (b *Local) opApply(
@@ -112,10 +118,20 @@ func (b *Local) opApply(
 	// stateHook uses schemas for when it periodically persists state to the
 	// persistent storage backend.
 	stateHook.Schemas = schemas
-	persistInterval := getEnvAsInt(persistIntervalEnvironmentVariableName, defaultPersistInterval)
+	persistInterval, intervalDiags := getEnvAsInt(persistIntervalEnvironmentVariableName, defaultPersistInterval)
+	diags = diags.Append(intervalDiags)
+	if intervalDiags.HasErrors() {
+		op.ReportResult(runningOp, diags)
+		return
+	}
 	if persistInterval < defaultPersistInterval {
-		panic(fmt.Sprintf("Can't use value lower than %d for env variable %s, got %d",
-			defaultPersistInterval, persistIntervalEnvironmentVariableName, persistInterval))
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Invalid environment variable value",
+			fmt.Sprintf("The environment variable %s must be at least %d, got %d.", persistIntervalEnvironmentVariableName, defaultPersistInterval, persistInterval),
+		))
+		op.ReportResult(runningOp, diags)
+		return
 	}
 	stateHook.PersistInterval = time.Duration(persistInterval) * time.Second
 

--- a/internal/backend/local/backend_apply.go
+++ b/internal/backend/local/backend_apply.go
@@ -42,7 +42,7 @@ func getEnvAsInt(envName string, defaultValue int) (int, tfdiags.Diagnostics) {
 			diags = diags.Append(tfdiags.Sourceless(
 				tfdiags.Error,
 				"Invalid environment variable value",
-				fmt.Sprintf("The environment variable %s must be a valid integer, got %q.", envName, val),
+				fmt.Sprintf("The environment variable %q is expected to be a valid integer but got %q which cannot be converted.", envName, val),
 			))
 			return 0, diags
 		}
@@ -128,7 +128,7 @@ func (b *Local) opApply(
 		diags = diags.Append(tfdiags.Sourceless(
 			tfdiags.Error,
 			"Invalid environment variable value",
-			fmt.Sprintf("The environment variable %s must be at least %d, got %d.", persistIntervalEnvironmentVariableName, defaultPersistInterval, persistInterval),
+			fmt.Sprintf("The value for the environment variable %q cannot be smaller than %d but got %d.", persistIntervalEnvironmentVariableName, defaultPersistInterval, persistInterval),
 		))
 		op.ReportResult(runningOp, diags)
 		return

--- a/internal/backend/local/backend_apply_test.go
+++ b/internal/backend/local/backend_apply_test.go
@@ -478,7 +478,6 @@ func TestLocal_applyInvalidPersistInterval(t *testing.T) {
 		TestLocalProvider(t, b, "test", applyFixtureSchema())
 
 		op, done := testOperationApply(t, "./testdata/apply")
-		defer done(t)
 
 		run, err := b.Operation(context.Background(), op)
 		if err != nil {
@@ -487,6 +486,9 @@ func TestLocal_applyInvalidPersistInterval(t *testing.T) {
 		<-run.Done()
 		if run.Result == backend.OperationSuccess {
 			t.Fatalf("expected operation to fail with invalid %s=abc", persistIntervalEnvironmentVariableName)
+		}
+		if got, want := done(t).Stderr(), "Invalid environment variable value"; !strings.Contains(got, want) {
+			t.Errorf("expected stderr to contain %q, got:\n%s", want, got)
 		}
 	})
 
@@ -497,7 +499,6 @@ func TestLocal_applyInvalidPersistInterval(t *testing.T) {
 		TestLocalProvider(t, b, "test", applyFixtureSchema())
 
 		op, done := testOperationApply(t, "./testdata/apply")
-		defer done(t)
 
 		run, err := b.Operation(context.Background(), op)
 		if err != nil {
@@ -506,6 +507,9 @@ func TestLocal_applyInvalidPersistInterval(t *testing.T) {
 		<-run.Done()
 		if run.Result == backend.OperationSuccess {
 			t.Fatalf("expected operation to fail with invalid %s=5", persistIntervalEnvironmentVariableName)
+		}
+		if got, want := done(t).Stderr(), "Invalid environment variable value"; !strings.Contains(got, want) {
+			t.Errorf("expected stderr to contain %q, got:\n%s", want, got)
 		}
 	})
 }

--- a/internal/backend/local/backend_apply_test.go
+++ b/internal/backend/local/backend_apply_test.go
@@ -428,3 +428,85 @@ func TestApply_applyCanceledAutoApprove(t *testing.T) {
 	}
 
 }
+
+func TestGetEnvAsInt(t *testing.T) {
+	const testEnv = "TEST_GET_ENV_AS_INT"
+
+	t.Run("env not set returns default", func(t *testing.T) {
+		os.Unsetenv(testEnv)
+		got, diags := getEnvAsInt(testEnv, 20)
+		if diags.HasErrors() {
+			t.Fatalf("unexpected error: %s", diags.Err())
+		}
+		if got != 20 {
+			t.Errorf("got %d, want 20", got)
+		}
+	})
+
+	t.Run("valid integer is parsed", func(t *testing.T) {
+		t.Setenv(testEnv, "30")
+		got, diags := getEnvAsInt(testEnv, 20)
+		if diags.HasErrors() {
+			t.Fatalf("unexpected error: %s", diags.Err())
+		}
+		if got != 30 {
+			t.Errorf("got %d, want 30", got)
+		}
+	})
+
+	t.Run("non-integer value returns error", func(t *testing.T) {
+		t.Setenv(testEnv, "abc")
+		_, diags := getEnvAsInt(testEnv, 20)
+		if !diags.HasErrors() {
+			t.Error("expected error but got none")
+		}
+	})
+
+	t.Run("float value returns error", func(t *testing.T) {
+		t.Setenv(testEnv, "1.5")
+		_, diags := getEnvAsInt(testEnv, 20)
+		if !diags.HasErrors() {
+			t.Error("expected error but got none")
+		}
+	})
+}
+
+func TestLocal_applyInvalidPersistInterval(t *testing.T) {
+	t.Run("non-integer value causes error diagnostic", func(t *testing.T) {
+		t.Setenv(persistIntervalEnvironmentVariableName, "abc")
+
+		b := TestLocal(t)
+		TestLocalProvider(t, b, "test", applyFixtureSchema())
+
+		op, done := testOperationApply(t, "./testdata/apply")
+		defer done(t)
+
+		run, err := b.Operation(context.Background(), op)
+		if err != nil {
+			t.Fatalf("unexpected error starting operation: %v", err)
+		}
+		<-run.Done()
+		if run.Result == backend.OperationSuccess {
+			t.Fatalf("expected operation to fail with invalid %s=abc", persistIntervalEnvironmentVariableName)
+		}
+	})
+
+	t.Run("below minimum value causes error diagnostic", func(t *testing.T) {
+		t.Setenv(persistIntervalEnvironmentVariableName, "5")
+
+		b := TestLocal(t)
+		TestLocalProvider(t, b, "test", applyFixtureSchema())
+
+		op, done := testOperationApply(t, "./testdata/apply")
+		defer done(t)
+
+		run, err := b.Operation(context.Background(), op)
+		if err != nil {
+			t.Fatalf("unexpected error starting operation: %v", err)
+		}
+		<-run.Done()
+		if run.Result == backend.OperationSuccess {
+			t.Fatalf("expected operation to fail with invalid %s=5", persistIntervalEnvironmentVariableName)
+		}
+	})
+}

--- a/internal/backend/local/backend_apply_test.go
+++ b/internal/backend/local/backend_apply_test.go
@@ -432,42 +432,62 @@ func TestApply_applyCanceledAutoApprove(t *testing.T) {
 func TestGetEnvAsInt(t *testing.T) {
 	const testEnv = "TEST_GET_ENV_AS_INT"
 
-	t.Run("env not set returns default", func(t *testing.T) {
-		got, diags := getEnvAsInt(testEnv, 20)
-		if diags.HasErrors() {
-			t.Fatalf("unexpected error: %s", diags.Err())
-		}
-		if got != 20 {
-			t.Errorf("got %d, want 20", got)
-		}
-	})
+	tests := []struct {
+		name         string
+		envValue     string
+		defaultValue int
+		wantValue    int
+		wantError    bool
+	}{
+		{
+			name:         "env not set returns default",
+			envValue:     "",
+			defaultValue: 20,
+			wantValue:    20,
+			wantError:    false,
+		},
+		{
+			name:         "valid integer is parsed",
+			envValue:     "30",
+			defaultValue: 20,
+			wantValue:    30,
+			wantError:    false,
+		},
+		{
+			name:         "non-integer value returns error",
+			envValue:     "abc",
+			defaultValue: 20,
+			wantError:    true,
+		},
+		{
+			name:         "float value returns error",
+			envValue:     "1.5",
+			defaultValue: 20,
+			wantError:    true,
+		},
+	}
 
-	t.Run("valid integer is parsed", func(t *testing.T) {
-		t.Setenv(testEnv, "30")
-		got, diags := getEnvAsInt(testEnv, 20)
-		if diags.HasErrors() {
-			t.Fatalf("unexpected error: %s", diags.Err())
-		}
-		if got != 30 {
-			t.Errorf("got %d, want 30", got)
-		}
-	})
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.envValue != "" {
+				t.Setenv(testEnv, tt.envValue)
+			}
 
-	t.Run("non-integer value returns error", func(t *testing.T) {
-		t.Setenv(testEnv, "abc")
-		_, diags := getEnvAsInt(testEnv, 20)
-		if !diags.HasErrors() {
-			t.Error("expected error but got none")
-		}
-	})
-
-	t.Run("float value returns error", func(t *testing.T) {
-		t.Setenv(testEnv, "1.5")
-		_, diags := getEnvAsInt(testEnv, 20)
-		if !diags.HasErrors() {
-			t.Error("expected error but got none")
-		}
-	})
+			got, diags := getEnvAsInt(testEnv, tt.defaultValue)
+			if tt.wantError {
+				if !diags.HasErrors() {
+					t.Errorf("expected error but got none, value=%d", got)
+				}
+				return
+			}
+			if diags.HasErrors() {
+				t.Fatalf("unexpected error: %s", diags.Err())
+			}
+			if got != tt.wantValue {
+				t.Errorf("got %d, want %d", got, tt.wantValue)
+			}
+		})
+	}
 }
 
 func TestLocal_applyInvalidPersistInterval(t *testing.T) {

--- a/internal/backend/local/backend_apply_test.go
+++ b/internal/backend/local/backend_apply_test.go
@@ -433,7 +433,6 @@ func TestGetEnvAsInt(t *testing.T) {
 	const testEnv = "TEST_GET_ENV_AS_INT"
 
 	t.Run("env not set returns default", func(t *testing.T) {
-		os.Unsetenv(testEnv)
 		got, diags := getEnvAsInt(testEnv, 20)
 		if diags.HasErrors() {
 			t.Fatalf("unexpected error: %s", diags.Err())

